tools/distpartitioning/globalids.py

```diff
@@ -0,0 +1,260 @@
+import numpy as np
+import torch
+import operator
+import itertools
+from gloo_wrapper import allgather_sizes, alltoall_cpu, alltoallv_cpu
+
+def get_shuffle_global_nids(rank, world_size, global_nids_ranks, parts_map, node_data):
+    """ For nodes which are not owned by the current rank, whose global_nid <-> shuffle_global-nid 
+    is not present at the current rank, this function retrieves their shuffle_global_ids from the owning
+    rank
+
+    Parameters: 
+    -----------
+    rank : integer
+        rank of the process
+    world_size : integer
+        total no. of ranks configured
+    global_nids_ranks : list
+        list of lists of global_nids, index of the list is the rank of the process
+                    where global_nid <-> shuffle_global_nid mapping is located. 
+    parts_map : dictionary
+        metis partitions, which are global_nid <-> rank mappings. 
+    node_data : numpy ndarray, integers
+        node data with additional columns inserted
+
+    Returns:
+    --------
+    list
+        list of shuffle_global_nids
+    list
+        list of global_nids whose shuffle_global_nids were retrieved from other ranks
+    """
+    #build a list of sizes (lengths of lists)
+    sizes = [len(x) for x in global_nids_ranks]
+
+    #compute total_nodes whose mappings should be resolved, between orig-node-id <-> global-id
+    total_nodes = np.sum(sizes)
+    if (total_nodes == 0):
+        print('Rank: ', rank, ' -- All mappings are present locally... No need for to send any info.')
+        return [], []
+
+    #determine the no. of orig-node-ids to send and receive and perform alltoall
+    send_counts = list(torch.Tensor(sizes).type(dtype=torch.int64).chunk(world_size))
+    recv_counts = list(torch.zeros([world_size], dtype=torch.int64).chunk(world_size))
+    alltoall_cpu(rank, world_size, recv_counts, send_counts)
```

Comment:
is `alltoall_cpu` used to send small data? if not, the inputs shouldn't be python lists.

Reply:
In this instance, alltoall_cpu is used to send sizes, which are just integers. And this function requires that the input and outputs to be in lists. Please note that the size of the list will not be greater than the world_size.